secret_store/src/key_server_cluster/cluster.rs

```diff
@@ -1219,40 +1242,46 @@ pub mod tests {
 
 		fn send(&self, to: &NodeId, message: Message) -> Result<(), Error> {
 			debug_assert!(&self.id != to);
-			self.data.lock().messages.push_back((to.clone(), message));
+			self.data.write().messages.push_back((to.clone(), message));
 			Ok(())
 		}
 
 		fn is_connected(&self, node: &NodeId) -> bool {
-			let data = self.data.lock();
+			let data = self.data.read();
 			&self.id == node || data.nodes.contains(node)
 		}
 
 		fn nodes(&self) -> BTreeSet<NodeId> {
-			self.data.lock().nodes.iter().cloned().collect()
+			self.data.read().nodes.iter().cloned().collect()
 		}
 
 		fn configured_nodes_count(&self) -> usize {
-			self.data.lock().nodes.len()
+			self.data.read().nodes.len()
 		}
 
 		fn connected_nodes_count(&self) -> usize {
-			self.data.lock().nodes.len()
+			self.data.read().nodes.len()
 		}
 	}
 
-	pub fn loop_until<F>(core: &mut Core, timeout: Duration, predicate: F) where F: Fn() -> bool {
+	pub fn loop_until<F>(runtime: &mut Runtime, timeout: Duration, predicate: F)
+		where F: Send + 'static + Fn() -> bool
+	{
 		let start = Instant::now();
-		loop {
-			core.turn(Some(Duration::from_millis(1)));
+
+		let check_predicate = future::loop_fn((), move|_| {
 			if predicate() {
-				break;
+				Ok(Loop::Break(()))
+			} else {
+				if Instant::now() - start > timeout {
+					panic!("no result in {:?}", timeout);
+				} else {
+					Ok(Loop::Continue(()))
+				}
 			}
+		}).map_err(|_: ()| ());
 
-			if Instant::now() - start > timeout {
-				panic!("no result in {:?}", timeout);
-			}
-		}
+		runtime.block_on(check_predicate).unwrap();
```

Comment:
afaiu, this will block all other futures execution if there's only one thread in `Runtime`' `ThreadPool` - `block_on` would just call `predicate()` until it timeouts. Probably worth specifying number of threads using `Builder` explicitly? Or just choose some other check mechanism.

Reply:
I've been able to fix test-hangs-issue, but still not sure about the reason :/ Probably polling never-ending LoopFn causes some changes in other futures poll schedule - could be investigated separately.
So the fix is:
1. replace your version of `loop_until` with:
```rust
	pub fn loop_until<F>(runtime: &mut Runtime, timeout: Duration, predicate: F)
		where F: Send + 'static + Fn() -> bool
	{
		use futures::Stream;
		use tokio::timer::Interval;

		let start = Instant::now();

		runtime.block_on(Interval::new_interval(Duration::new(0, 1_000_000))
			.and_then(move |_| {
				if Instant::now() - start > timeout {
					panic!("no result in {:?}", timeout);
				}

				Ok(())
			})
			.take_while(move |_| future::ok(!predicate()))
			.for_each(|_| Ok(()))
			.then(|_| future::ok::<(), ()>(()))
		).unwrap();
	}
```
2. increase TIMEOUT constant to 1s (i.e. `const TIMEOUT: Duration = Duration::from_millis(1000);`) - probably new implementation of `loop_until` causes multiple thread-context switches because I got timeout several times in simplest `cluster_connects_to_other_nodes` test, even though it has been near-to-completion.

Ran `while reset && cargo test -p ethcore-secretstore --features secretstore; do :; done` for last hour without any issues (except single one - unrelated, will open a PR later today).

Reply:
Ok looks great. I'll make those adjustments.

Just to clarify, `Runtime::block_on` is basically equivalent to `Core::run`, `tokio::run` etc. It basically starts the runtime up. It should never be called from within a future, etc. so it wouldn't actually block anything else from happening, it just blocks the containing thread (which, again, should never be a future/task).

I'll explicitly set the thread pool count to 4 (which is a good round number to use in general) but if there are problems created by using other counts (like 1) it usually indicates some deeper problem (one of the futures is blocking the entire thread waiting on a mutex, etc.). As an aside, I have a crate, [Qutex](https://github.com/cogciprocate/qutex), which provides non-thread-blocking futures-aware mutexes which can be useful when dealing with IO and shared data.